### DISCUSSION TITLE
How to use cosmopolitan c as a libc replacement outside the cosmo ecosystem, with regular updates?

### DISCUSSION THREAD

**Discussion by vinodkd** [jart/cosmopolitan#1341] [Q&A] [answered]
How do I use cosmopolitan c as a replacement for libc outside the cosmo ecoysystem (ie, purely as a dependent library), while also keeping up with changes to it? 

Long time ago, I used the [original APE blog post](https://justine.lol/ape.html) to download the binaries and setup make to use them, but that was point-in-time and I'm still stuck with that version. Recently, I found [this post on the internet](https://jeskin.net/blog/getting-started-with-cosmopolitan-libc/) which goes one step better by pulling the binaries (as long as a checksum matches) from within the make file.

Is this the best way to do it? How can I do the equivalent of "get latest" every time I build my code?

Note: I know there are benefits to being inside the cosmo ecosystem, but it feels like an overkill for my small project, hence this question.

**Comment by jart** [comment, marked as answer]
The flagship example for doing this is llamafile, which sets up a monorepo that's built externally by cosmocc. Feel free to use its boilerplate:

- https://github.com/Mozilla-Ocho/llamafile/blob/main/build/download-cosmocc.sh
- https://github.com/Mozilla-Ocho/llamafile/blob/main/build/config.mk
- https://github.com/Mozilla-Ocho/llamafile/blob/main/build/deps.mk
- https://github.com/Mozilla-Ocho/llamafile/blob/main/build/rules.mk
- https://github.com/Mozilla-Ocho/llamafile/blob/main/build/sha256sum.c
- https://github.com/Mozilla-Ocho/llamafile/blob/main/Makefile

The basic idea is you get all your make rules to depend on this target that downloads your toolchain.

https://github.com/Mozilla-Ocho/llamafile/blob/a8fd4d28c3d2259c98af7035bcdda1a68af6f62c/build/config.mk#L55-L56

That target is defined as `$(COSMOCC)` and you can see the wildcard recipes depend on it, so compiles trigger the toolchain download:

https://github.com/Mozilla-Ocho/llamafile/blob/a8fd4d28c3d2259c98af7035bcdda1a68af6f62c/build/rules.mk#L8-L10

Your cosmocc toolchain is hosted redundantly via both our GitHub and our cosmo.zip web server, so it's guaranteed reliable for your CI. The downloader script is designed so that it'll failover if one of the servers is down or blocked by your firewall. The versioned artifacts for cosmo also never change, so each time you want to upgrade you just pin the new version with its checksum. I promise to maintain these URLs until the heat death of the universe in service to your projects.

**Reply by vinodkd** [reply]
Wow, thanks for the detailed and thoughtful response. I'll probably do a less robust, but simpler version of this for my project.

**Reply by vinodkd** [reply]
I should also add that my original make file was from when cosmocc didnt exist, so it used gcc and passed it the flags from your original ape blog post. Now that cosmocc does exist, i'll switch to using that since it's more self contained and it produces ape binaries by default.

**Comment by jart** [comment]
If you use Bazel then ARM maintains the code for integrating it with Cosmopolitan here: https://gitlab.arm.com/bazel/ape

**Comment by jart** [comment]
For examples of how to get other people's normative open source code to build using cosmocc, see https://justine.lol/cosmo3/#install which gives a few examples. For example, for projects with a simple Makefile like https://github.com/jart/json.cpp it is oftentimes sufficient to say:

```
make CC=cosmocc CXX=cosmoc++
```

For projects that use GNU Autotools, you'll often say something like:

```
export CC="cosmocc -I/opt/cosmos/include -L/opt/cosmos/lib"
export CXX="cosmoc++ -I/opt/cosmos/include -L/opt/cosmos/lib"
export PKG_CONFIG="pkg-config --with-path=/opt/cosmos/lib/pkgconfig"
export INSTALL="cosmoinstall"
export AR="cosmoar"
./configure --prefix=/opt/cosmos
make -j
make install
```

For other people's projects that use cmake, what I might say is:

```
mkdir -p cosmo
cd cosmo
cmake -S ../clang -B . \
-D CMAKE_C_COMPILER=cosmocc \
-D CMAKE_CXX_COMPILER=cosmoc++ \
-D CMAKE_AR="$(command -v cosmoar)" \
-D Python3_EXECUTABLE=/opt/cosmos/bin/python \
-D LLVM_ENABLE_PIC=OFF \
-D CMAKE_SKIP_RPATH=ON \
-D LLVM_BUILD_STATIC=OFF \
-D BUILD_SHARED_LIBS=OFF \
-D LLVM_LIT_ARGS=-v
```

For our own projects that use cmake, Cristian Adam from Qt team recommends something like:

```
set(CMAKE_SYSTEM_NAME Linux)

set(arch x86_64-unknown-cosmo)
set(_CMAKE_TOOLCHAIN_PREFIX ${arch}-)
set(CMAKE_C_COMPILER $ENV{COSMOCC}/bin/${arch}-cc)
set(CMAKE_CXX_COMPILER $ENV{COSMOCC}/bin/${arch}-c++)

set(CMAKE_CXX_FLAGS_INIT "-fexceptions")

set(CMAKE_AR $ENV{COSMOCC}/bin/${arch}-ar)
```

Finally let's say you want to go really advanced with @ahgamut level build mastery. For complicated projects like Emacs or GCC you might need to switch from using cosmocc (our all-in-one frontend wrapper script) to using the architecture-specific toolchains separately, creating two separate binaries, and then unifying them using the `apelink` program. The https://github.com/ahgamut/superconfigure/ project has examples of how we build such projects. It's actually what we use to build cosmocc. But if you want to simply get started with *understanding* how all the various programs in cosmocc work together, you need only ask `cosmocc` to log its underlying commands:

```
master jart@luna:~/cosmo$ export BUILDLOG=/tmp/log
master jart@luna:~/cosmo$ cosmocc -o /tmp/hello ~/scratch/hello.c
master jart@luna:~/cosmo$ cat /tmp/log
# /opt/cosmocc/bin/cosmocc -o /tmp/hello /home/jart/scratch/hello.c
(cd /home/jart/cosmo; /opt/cosmocc/bin/aarch64-linux-cosmo-gcc -o/tmp/fatcosmocc.t917bddd332f2.o -D__COSMOPOLITAN__ -D__COSMOCC__ -D__FATCOSMOCC__ -include libc/integral/normalize.inc -fportcosmo -fno-semantic-interposition -fno-optimize-sibling-calls -mno-omit-leaf-frame-pointer -fno-schedule-insns2 -Wno-implicit-int -ffixed-x18 -ffixed-x28 -fpatchable-function-entry=7,6 -fno-inline-functions-called-once -DFTRACE -DSYSDEBUG -fno-pie -nostdinc -isystem /opt/cosmocc/bin/../include -fsigned-char -c /home/jart/scratch/hello.c -fno-omit-frame-pointer)
(cd /home/jart/cosmo; /opt/cosmocc/bin/x86_64-linux-cosmo-gcc -o/tmp/fatcosmocc.but0r43q84na1.o -D__COSMOPOLITAN__ -D__COSMOCC__ -D__FATCOSMOCC__ -include libc/integral/normalize.inc -fportcosmo -fno-semantic-interposition -fno-optimize-sibling-calls -mno-omit-leaf-frame-pointer -fno-schedule-insns2 -Wno-implicit-int -mno-tls-direct-seg-refs -fpatchable-function-entry=18,16 -fno-inline-functions-called-once -DFTRACE -DSYSDEBUG -fno-pie -nostdinc -isystem /opt/cosmocc/bin/../include -c /home/jart/scratch/hello.c -fno-omit-frame-pointer)
(cd /home/jart/cosmo; /opt/cosmocc/bin/fixupobj /tmp/fatcosmocc.but0r43q84na1.o)
(cd /home/jart/cosmo; /opt/cosmocc/bin/fixupobj /tmp/fatcosmocc.t917bddd332f2.o)
(cd /home/jart/cosmo; /opt/cosmocc/bin/x86_64-linux-cosmo-gcc -o/tmp/fatcosmocc.2w68kd3okig83.com.dbg /opt/cosmocc/bin/../x86_64-linux-cosmo/lib/ape.o /opt/cosmocc/bin/../x86_64-linux-cosmo/lib/crt.o -static -nostdlib -no-pie -fuse-ld=bfd -Wl,-z,noexecstack -Wl,-z,norelro -Wl,--gc-sections -L/opt/cosmocc/bin/../x86_64-linux-cosmo/lib -L/opt/cosmocc/bin/../x86_64-linux-cosmo/lib -Wl,-T,/opt/cosmocc/bin/../x86_64-linux-cosmo/lib/ape.lds -Wl,-z,common-page-size=4096 -Wl,-z,max-page-size=16384 /tmp/fatcosmocc.but0r43q84na1.o -lcosmo)
(cd /home/jart/cosmo; /opt/cosmocc/bin/aarch64-linux-cosmo-gcc -o/tmp/fatcosmocc.8m4hoxt8pyt01.aarch64.elf /opt/cosmocc/bin/../aarch64-linux-cosmo/lib/crt.o -static -nostdlib -no-pie -fuse-ld=bfd -Wl,-z,noexecstack -Wl,-z,norelro -Wl,--gc-sections -L/opt/cosmocc/bin/../aarch64-linux-cosmo/lib -L/opt/cosmocc/bin/../aarch64-linux-cosmo/lib -Wl,-T,/opt/cosmocc/bin/../aarch64-linux-cosmo/lib/aarch64.lds -Wl,-z,common-page-size=16384 -Wl,-z,max-page-size=16384 /tmp/fatcosmocc.t917bddd332f2.o -lcosmo)
(cd /home/jart/cosmo; /opt/cosmocc/bin/fixupobj /tmp/fatcosmocc.8m4hoxt8pyt01.aarch64.elf)
(cd /home/jart/cosmo; /opt/cosmocc/bin/fixupobj /tmp/fatcosmocc.2w68kd3okig83.com.dbg)
(cd /home/jart/cosmo; /opt/cosmocc/bin/apelink -V -1 -l /opt/cosmocc/bin/ape-x86_64.elf -l /opt/cosmocc/bin/ape-aarch64.elf -M /opt/cosmocc/bin/ape-m1.c -o /tmp/hello /tmp/fatcosmocc.2w68kd3okig83.com.dbg /tmp/fatcosmocc.8m4hoxt8pyt01.aarch64.elf)
(cd /home/jart/cosmo; /opt/cosmocc/bin/pecheck /tmp/hello)
```
